test/integration/dualstack/dualstack_test.go

```diff
@@ -0,0 +1,1336 @@
+		},
+		{
+			name:               "Type ClusterIP - Server Allocated IP - IPv6 IPv4 Family - Policy Single Stack",
+			serviceType:        v1.ServiceTypeClusterIP,
+			clusterIPs:         []string{},
+			ipFamilies:         []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			expectedIPFamilies: []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			ipFamilyPolicy:     v1.IPFamilyPolicySingleStack,
+			expectError:        true,
+		},
+		{
+			name:               "Type ClusterIP - Server Allocated IP - IPv6 IPv4 Family - Policy Prefer Dual Stack",
+			serviceType:        v1.ServiceTypeClusterIP,
+			clusterIPs:         []string{},
+			ipFamilies:         []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			expectedIPFamilies: []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			ipFamilyPolicy:     v1.IPFamilyPolicyPreferDualStack,
+			expectError:        false,
+		},
+		{
+			name:               "Type ClusterIP - Server Allocated IP - IPv6 IPv4 Family - Policy Required Dual Stack",
+			serviceType:        v1.ServiceTypeClusterIP,
+			clusterIPs:         []string{},
+			ipFamilies:         []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			expectedIPFamilies: []v1.IPFamily{v1.IPv6Protocol, v1.IPv4Protocol},
+			ipFamilyPolicy:     v1.IPFamilyPolicyRequireDualStack,
+			expectError:        false,
+		},
+	}
+
+	for i, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+
+			svc := &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("svc-test-%d", i), // use different services for each test
+				},
+				Spec: v1.ServiceSpec{
+					Type:           tc.serviceType,
+					ClusterIPs:     tc.clusterIPs,
+					IPFamilies:     tc.ipFamilies,
+					IPFamilyPolicy: &tc.ipFamilyPolicy,
+					Ports: []v1.ServicePort{
+						{
+							Name:       fmt.Sprintf("port-test-%d", i),
+							Port:       443,
+							TargetPort: intstr.IntOrString{IntVal: 443},
+							Protocol:   "TCP",
+						},
+					},
+				},
+			}
+
+			// create a service
+			_, err = client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
+			if (err != nil) != tc.expectError {
+				t.Errorf("Test failed expected result: %v received %v ", tc.expectError, err)
+			}
+			// if no error was expected validate the service otherwise return
+			if err != nil {
+				return
+			}
+			// validate the service was created correctly if it was not expected to fail
+			svc, err = client.CoreV1().Services(metav1.NamespaceDefault).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Errorf("Unexpected error to get the service %s %v", svc.Name, err)
+			}
+
+			if err := validateServiceAndClusterIPFamily(svc, tc.expectedIPFamilies); err != nil {
+				t.Errorf("Unexpected error validating the service %s %v", svc.Name, err)
+			}
+		})
+	}
+}
+
+// TestUpgradeDowngrade tests upgrading and downgrading a service from/to dual-stack
+func TestUpgradeDowngrade(t *testing.T) {
```

Comment:
should we have an equivalent test against a single-stack cluster to prove you can't upgrade?

Reply:
That would be a validation test. There are plenty of them in other funcs in same file.